### PR TITLE
Repurpose Gen0ReductionCount field for dynamic heap count diagnostics.

### PR DESCRIPTION
**Pull request by PeterSolMS**
A quick hack to make diagnostics for dynamic heap count easier.

### CODE REVIEW

**Comment by Maoni0** [comment]
sorry I wasn't being clear - I never meant that we were going to check in the repurpose part - it was just there for testing purposes. so we wouldn't need to wait for the parsing work to be done _to look at the traces_. the parsing work isn't high priority for testing since we could easily find some fields to repurpose.

the rest of this PR is obviously needed and should be included in the build for testing purposes. @PeterSolMS do you mind creating another PR "adaptation fixes and improvements" that encapsulates the hang fix and the msl overhead instability/finalizer lock improvement?

File: src/coreclr/gc/gc.cpp [hunk 1]
```diff
@@ -1413,7 +1413,7 @@ bool gc_heap::should_move_heap (GCSpinLock* msl)
 
 // All the places where we could be stopped because there was a suspension should call should_move_heap to check if we need to return
 // so we can try another heap or we can continue the allocation on the same heap.
-enter_msl_status gc_heap::enter_spin_lock_msl (GCSpinLock* msl)
+enter_msl_status gc_heap::enter_spin_lock_msl_helper (GCSpinLock* msl)
```
Comment (Maoni0) [inline src/coreclr/gc/gc.cpp:1416]: can we change this method a bit so we don't immediately need to do another Interlocked check? you could change it to a do/while loop. 

also we should run some microbenchmarks to see if this regresses anything (especially WKS GC)...

**Review by Maoni0 — APPROVED** [review]
LGTM!

**Comment by Maoni0** [comment]
Actually, talking to Moko just now made me realize that it’s probably better if we make it perfectly obviously if adaption code is there but adaptation is not turned on so maybe in that case we just make the repurposed builds have -1 for those fields?

**Comment by Maoni0** [comment]
hmm, I'm not sure what happened? I wanted to push some more (temporary) instrumentation changes to this PR but it instantly triggered some sort wide code review requests... sorry about that! please ignore!

**Comment by mangod9** [comment]
@cshung @mrsharm assume we could close this since we plan to enable dynamic events?

**Comment by mrsharm** [comment]
This PR consists of instrumentation and fixes. The fixes have been merged with the following PR: https://github.com/dotnet/runtime/pull/87218/files - we should close this PR and extract the instrumentation logic.

**Pull request closed without merging**